docs/fsharp/style-guide/component-design-guidelines.md

```diff
@@ -693,11 +693,83 @@ let checkNonNull argName (arg: obj) =
     | null -> nullArg argName
     | _ -> ()
 
-let checkNonNull` argName (arg: obj) =
+let checkNonNull' argName (arg: obj) =
     if isNull arg then nullArg argName
     else ()
 ```
 
+Starting with F# 9, you can leverage the new `| null` [syntax](../language-reference/values/null-values.md#null-values-starting-with-f-9) to make the compiler indicate possible null values and where they need handling. The above code will then start producing warnings:
+
+```fsharp
+let checkNonNull argName (arg: obj) =
+    match arg with
+    // nullness warning: the type 'obj' does not support 'null'
+    | null -> nullArg argName
+    | _ -> ()
+
+let checkNonNull' argName (arg: obj) =
+    // nullness warning: the type 'obj' does not support 'null'
+    if isNull arg then nullArg argName 
+    else ()
+```
+
+To address the warnings, you could explicitly specify `null` as a possible argument value:
+
+```fsharp
+let checkNonNull argName (arg: obj | null) =
+    match arg with
+    | null -> nullArg argName
+    | _ -> ()
+
+let checkNonNull' argName (arg: obj | null) =
+    if isNull arg then nullArg argName 
+    else ()
+```
+
+Yet, the most succinct way to achieve this is to use the new builtin `nullArgCheck` function:
```

Comment:
This does not fit here, `nullArgCheck` is a throwing validation, more of a last resort.
The motivating examples should start be doing the right thing by handling nulls without failing, this should not be demonstrated as a succint way.

Reply:
Fair, `nullArgCheck` is mentioned elsewhere and doesn't need to be here.